## title: First Project

**Description**
- Must be free apps, revenue only in apps ads.
- Remove non-english apps
- Delete inaccurate data
- Detect duplicate data, and remove it
- Revenue is based on number of app users 

**Goal**
- analyze data, help devs understanding type of apps are likely attract ppl+ 

In [16]:
# import modules 
from csv import reader

In [17]:
# select dataset to be readed
def open_dataset(dataset):
    opened_file = open(dataset, encoding = 'utf8')
    read_file = reader(opened_file)
    data = list(read_file)
    data_header = data[0]
    return data

# datasets:
dataset_1 = 'datasets\AppleStore.csv' # Apple dataset
dataset_2 = 'datasets\googleplaystore.csv' # Google Play dataset

# asign variable to function
ios_set = open_dataset(dataset_1)
androidSet = open_dataset(dataset_2)


### Explore Dataset...
In order to explore the dataset we create the function <i>exploredata()</i> and ask for four parameters; the dataset, start index number, ending index number and a boolean data type to give the user the option to display total number of rows and column in dataset

In [18]:
# create function to read the dataset
def explore_data(dataset, start, end, rows_and_cols= False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n') # add empty line after each iteration
        
    if rows_and_cols:
        print(f'Number of rows: {len(dataset)}')
        print(f'Number of Columns: {len(dataset[0])}')
        
    


In [175]:
# call function for Apple apps Dataset
print('Sample of Apple dataset:')
print(explore_data(dataset= ios_set, start= 0, end= 3, rows_and_cols=True))
# print(apps_data)

Sample of Apple dataset:
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 7198
Number of Columns: 16
None


In [178]:
# call function for Google apps dataset
print('Sample of Android dataset:')
print(explore_data(dataset= androidSet, start= 0, end= 3, rows_and_cols=True))
# print dataset
# print(apps_data)

Sample of Android dataset:
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10841
Number of Columns: 13
None


### Data cleaning
- Find the row/s that has a missing a value
    - The Google play data set(androidSet) has a dedicated [dicussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion) and we can see that [one discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) outlines an error for row 10473. lets check that out 
- Eliminate duplicate apps entries

**SearchCol()**: <br>
This function will track duplicates and any data points that are in blank

In [283]:
def searchCol(dataset, idx, searchItem, start= 0, end=None):
    if end == None:
        end == len(dataset)
        
    for app in dataset[start:end]:
        if app[idx] == searchItem:
            print(app)

Below we search for spaces using searchCol(). We searched the app name column for Instagram and you can see that there are duplicates for this app. We can also see that there are two apps with empty data points in the "Content Rating" and "Current Ver" column. Additionally the app named "Life Made WI-Fi Touchscreen Photo Frame" has "Everyone" under the "Price" column when it should be a number displayed as a string. We will address this in the Data Cleaning section.

In [284]:
print(androidSet[0])
searchCol(androidSet, 0, 'Instagram')
searchCol(androidSet, 8, "")
searchCol(androidSet, 11, "")

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Market Update Helper', 'LIBRARIES_AND_DEMO', '4.1', '20145', '11k', '1,000,000+', 'Free', '0', 'Everyone', 'Libraries & Demo', 'February 12, 2013'

Now we explore the Apple dataset. We search for two apps that we knew were duplicates by looking at the documentation for the dataset.

In [285]:
print(ios_set[0])
searchCol(ios_set, 1, "Mannequin Challenge")
searchCol(ios_set, 1, "VR Roller Coaster")

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['1173990889', 'Mannequin Challenge', '109705216', 'USD', '0.0', '668', '87', '3.0', '3.0', '1.4', '9+', 'Games', '37', '4', '1', '1']
['1178454060', 'Mannequin Challenge', '59572224', 'USD', '0.0', '105', '58', '4.0', '4.5', '1.0.1', '4+', 'Games', '38', '5', '1', '1']
['952877179', 'VR Roller Coaster', '169523200', 'USD', '0.0', '107', '102', '3.5', '3.5', '2.0.0', '4+', 'Games', '37', '5', '1', '1']
['1089824278', 'VR Roller Coaster', '240964608', 'USD', '0.0', '67', '44', '3.5', '4.0', '0.81', '4+', 'Games', '38', '0', '1', '1']


## Data Clearing
___
We must clean the data before we can analyze it to find characteristics of X app. We mentioned hoe the Google dataset has some apps with missing datapoints. Missing data is the first item we'll address in the dataset cleanup 

In [21]:
# # iterate through the dataset(androidSet or ios_set) to find the row that is missing a value
# for row in androidSet:
#     if len(row) != len(androidSet[0]):
#         print(row)
#         print(androidSet.index(row))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10473


- If there is a print from ios_set run next cell
- if there is a print from androidSet run next cell
    - Row index 10473 was print: app's name 'Life Made WI-Fi Touchscreen Photo Frame', the rating is 19. This is clearly off because maximum rating for a Google Play app is 5. We'll delete this row. 

In [22]:
# remove row founded
print(f'Android set lenght before deleting row 10473: {len(androidSet)}')
del androidSet[10473] # run this cell only onces
print(f'Android set lenght after deleting row 10473: {len(androidSet)}')

Android set lenght before deleting row 10473: 10842
Android set lenght after deleting row 10473: 10841


### Remove Duplicate Entries
**Part One**
- Explore google play dataset long enough, to find that there're some apps that hace more that one entry. For instance, Instagram has 4 entries 

In [23]:
# Case: AndroidSet - lets review one of the duplicated apps in this case: Instagram
# Case: ios_set - lets review one of the duplicated apps in this case: VR Rollet Coaster

for app in ios_set: # change dataset if required
    name = app[1]
    if name == 'VR Roller Coaster': # change app's name if required
        print(app)

['952877179', 'VR Roller Coaster', '169523200', 'USD', '0.0', '107', '102', '3.5', '3.5', '2.0.0', '4+', 'Games', '37', '5', '1', '1']
['1089824278', 'VR Roller Coaster', '240964608', 'USD', '0.0', '67', '44', '3.5', '4.0', '0.81', '4+', 'Games', '38', '0', '1', '1']


In [289]:
# find duplicate apps in ios_set

def dupleCount(dataset, idx):
    # create empty lists
    duplicate_apps = []
    unique_apps = []

    for app in dataset:
        # if is for androidSet change to idx 0 for name
#         name = app[1]
        if app[idx] in unique_apps:
            duplicate_apps.append(app[idx])
        else:
            unique_apps.append(app[idx])
    # print the follow info
    print(f'Unique Apps: {len(unique_apps)}')
    print(f'Duplicate Apps: {len(duplicate_apps)}')
    print(f'Total Apps: {len(duplicate_apps) + len(unique_apps)}')
    print(f'Some Duplicate Apps:\n {duplicate_apps[:15]}')
    

dupleCount(ios_set, 1)
print('\n')
dupleCount(androidSet, 0)

Unique Apps: 7196
Duplicate Apps: 2
Total Apps: 7198
Some Duplicate Apps:
 ['Mannequin Challenge', 'VR Roller Coaster']


Unique Apps: 9660
Duplicate Apps: 1181
Total Apps: 10841
Some Duplicate Apps:
 ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


### PART 2...

We will: <br> 
- create dict, where each key is unique app name, and **value is the highest number** of that app
- Use dict to create new dataset, which will have **only one entry per app** (<i> we only select apps with the highest number of reviews<i>)

In [215]:
# Leave duplicated app w/ higher reviews number, remove the others from dataset; androidandroidSet that contains 

# create empty dict
ios_review_max = {}


for app in ios_set[1:]: # change dataset as required
    # if is for androidSet change idx to 0 for name and to idx 3 for n_reviews
    name = app[1] 
    n_reviews = app[5]
    # if exist in...
    if name in ios_review_max and ios_review_max[name] < n_reviews:
        ios_review_max[name] = n_reviews
    # elif not...
    elif name not in ios_review_max:
        ios_review_max[name] = n_reviews

# Check that entries in dict still 7195 as unique_apps list
print(f'ios Expected lenght: {len(ios_set[1:]) - 2}')
print(f'ios Actual lenght: {len(ios_review_max)}')

ios Expected lenght: 7195
ios Actual lenght: 7195


In [211]:
# Leave duplicated app w/ higher reviews number, remove the others from dataset; androidandroidSet that contains 

# create empty dict
android_review_max = {}


for app in androidSet[1:]: # change dataset as required
    # if is for androidSet change idx to 0 for name and to idx 3 for n_reviews
    name = app[0] 
    n_reviews = app[3]
    # if exist in...
    if name in android_review_max and android_review_max[name] < n_reviews:
        android_review_max[name] = n_reviews
    # elif not...
    elif name not in android_review_max:
        android_review_max[name] = n_reviews

# Check that entries in dict still 9661 as unique_apps list (9660 + header row)
print(f'android Expected lenght: {len(androidSet[1:]) - 1181}')
print(f'android Actual lenght: {len(android_review_max)}' )

android Expected lenght: 9659
android Actual lenght: 9659


Lets use the <i>reviews_max</i> dict to remove duplicates. For duplicate cases, we'll only keep entries of highest number of reviews
- initialized two empty lists: <i>android_clean</i> & <i>already_added</i>
- Iterate through <i>androidSet</i> dataset, and for every iteration:
    - Isolate name of app & number of reviews
    - Add current row(app) to <i>android_clean</i> list & app name (<i>name</i>) to <i>already_added</i> list if:
        - number of reviews of current app matches number of reviews of that app as described in review_max dict, and
        - name of app is not already in <i>already_added</i> list. We need add this suplementary condition to account for cases where highest number of reviews of duplicate app is the same for more than one entry. If we check only for **review_max[name] == n_reviews**, we will still end up with duplicates entries for apps. 

In [213]:
# removing....
ios_clean = []
already_added= [] # will store just apps names



for app in ios_set[1:]:
    # if is for androidSet change idx to 0 for name and to idx 3 for n_reviews
    name = app[1]
    n_reviews = app[5]

    if (ios_review_max[name] == n_reviews ) and (name not in already_added):
        ios_clean.append(app)
        already_added.append(name) # make sure this is inside of the if block


In [216]:
# Now let's explore the new data set, and confirm that th number of ios rows is 7195
explore_data(ios_clean, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7195
Number of Columns: 16


In [214]:
# removing....
android_clean= [] # will store new clean dataset
already_added= [] # will store just apps name


for app in androidSet[1:]:
    # if is for androidSet change idx to 0 for name and to idx 3 for n_reviews
    name = app[0]
    n_reviews = app[3]

    if (android_review_max[name] == n_reviews ) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name) # make sure this is inside of the if block


In [217]:
# Now let's explore the new data set, and confirm that th number of android rows is 9659
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of Columns: 13


### Check if android and ios datasets are now free of duplicates
Lets print the lenght of android_clean and ios_clean

In [218]:
# MAKE SURE THAT PART ONE AND TWO FOR REMOVING DUPLICATES RAN FOR BOTH DATASETS
# BEFORE RUNNING THIS CELL

# ios_clean lenght
print(f'Lenght of ios_clean: {len(ios_clean)}')
# andriod_clean
print(f'Lenght of Android_clean: {len(android_clean)}')

Lenght of ios_clean: 7195
Lenght of Android_clean: 9659


## Removing Non-English Apps
**Part One**
- Because we are not interested to keep apps that are not english-oriented audience, we'll remove them, One way is to remove each app that contains non-english symbols -English text books usually included letters from english alphabet, numbers 0-9, punctuation marks  (., !, ?, ;, etc.), and other symbols (+, *, /, etc.).
- All these characters english only, are encoded using USCII standard. Each USCII character correspond to a number between 0-127 associated with it.
- Lets build a function that check each name weather it contains non-ASCII characters.

- the function below, we will used **ord()** function to finf out the corresponding encondig number of each character

In [109]:
# evaluate if is a common english character
def is_english(string):
    for character in string:
        if ord(character) > 127:
            return False
        
    return True

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(ord('™'))
print(ord('😜'))

True
False
False
False
8482
128540


The function work fine, but some english apps names use emojis or other symbols (™, — (em dash), – (en dash), etc.). That fall out of the ASCII range, for this reason we need to modify the function that have above.

### Part Two
to minimized the impact we'll remove only an app that has more than three non-ASCII character

In [116]:
def is_english(string):
    # add counter for non-ASCII characters
    non_ASCII = 0
    # Iterate through string
    for character in string:
        if ord(character) > 127:
            #Add non_ASCII to variable
            non_ASCII +=1
        # if more than 3 non-ASCII return False
        if non_ASCII >3:
            return False
        
    return True

# call the function
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Instachat 😜'))

True
False
True


In [219]:
# Create 2 list that will contain only english apps, do it for both datasets
ios_english = []
android_english = []


# Iterate through datasets

for app in ios_clean:
    name = app[1] # idx 1 is the column that contains the name
    if is_english(name):
        ios_english.append(app)

        
for app in android_clean:
    name = app[0] # idx 1 is the column that contains the name
    if is_english(name):
        android_english.append(app)# append app, because you want to add the entire row


### Let's check the new shape of the datasets
These datasets will only contain english apps

In [220]:
# Lets check the new size of the ios_english dataset
explore_data(ios_english, 0, 3, True)

# Lets check the new size of the ios_english dataset
explore_data(android_english, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 6181
Number of Columns: 16
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies

## Isolating the Free Apps
we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps, and we'll need to isolate only the free apps for our analysis. Below, we isolate the free apps for both our data sets.

In [225]:
# create lists for the datasets that willl only contain free apps
ios_final = []
android_final = []

# Iterate through the datasets to only append the new list apps that are 'Free'
for app in ios_english:
    price = app[4] # idx 4 is the the price column in IOS
    if price == '0.0':
        ios_final.append(app)
        
for app in android_english:
    price = app[7]
    if price == '0': # idx 7 is the the price column in Android
        android_final.append(app)

In [228]:
# check the lenght of each one of the final datasets
print(f'We can see that we have left with {len(ios_final)} IOS apps and {len(android_final)} Android apps')

We can see that we have left with 3220 IOS apps and 8862 Android apps


## Most Common Apps by Genre
### Part One
our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

    1.Build a minimal Android version of the app, and add it to Google Play.
    2.If the app has a good response from users, we then develop it further.
    3.If the app is profitable after six months, we also build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of the most common genres for each market. For this, we'll build a frequency table for the <i>prime_genre</i> column of the App Store data set, and the <i>Genres</i> and <i>Category</i> columns of the Google Play data set.

### Part Two
We'll build two functions we can use to analyze the frequency tables:

- One function to generate frequency tables that show percentages
- Another function that we can use to display the percentages in a descending order


In [231]:
# create function called freq_table()
def freq_table(dataset, idx):
    # create a dictionary and list
    table = {}
    total = 0
    
    # This for LOOP will create a frequency table
    for app in dataset[1:]:
        total += 1
        value = app[idx]
#         if column, then...
        if value in table:
            table[value] +=1
#         if not, then...
        else:
            table[value] = 1
            
            
    # create dict that will contain the percentage 
    table_percentage = {}
    # iterate through table to get percentage value
    for key in table:
        percentage = (table[key]/ total) *100
        # Use the table[key] as key for table_percentage, then add % as its value 
        table_percentage[key] = percentage
        
    return table_percentage
            
# create function that will create a frequency table from any column provided using 
# freq_table() function, 
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    # Iterate through table created
    for key in table:
        # convert table into list of tuples
        key_val_as_tuple = (table[key], key)
        # append the tuple lists to the created table_display list
        table_display.append(key_val_as_tuple)
        
    # Sorts the list in a descending order
    table_sorted = sorted(table_display, reverse = True)
    # Iterate thorugh the list to prints its entries in descending order
    for entry in table_sorted:
        print(entry[1], ':', entry[0]) # idx 1=> genre , idx=> 0 is percentage

### Part Three
Let's examinate the frequency table for the <i>prime_genre</i> column of IOS App store dataset

In [232]:
# Prime Genre
display_table(ios_final, -5 )

Games : 58.1547064305685
Entertainment : 7.890649269959615
Photo & Video : 4.970487729108418
Education : 3.6657347002174587
Social Networking : 3.2618825722273996
Shopping : 2.60950605778192
Utilities : 2.516309412861137
Sports : 2.1435228331780056
Music : 2.050326188257223
Health & Fitness : 2.019260639950295
Productivity : 1.7396707051879468
Lifestyle : 1.5843429636533086
News : 1.3358185771978874
Travel : 1.2426219322771046
Finance : 1.1183597390493942
Weather : 0.8698353525939734
Food & Drink : 0.8077042559801181
Reference : 0.5591798695246971
Business : 0.5281143212177695
Book : 0.4349176762969867
Navigation : 0.1863932898415657
Medical : 0.1863932898415657
Catalogs : 0.12426219322771047


We can see that among the free English apps more than half (58.16%) are games. Entertaiment apps are close to (8%), followed by photos and videos apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.<br>

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.<br>

Let's continue by examining the <i>Genres</i> and <i>Category</i> columns of the Google Play data set (two columns which seem to be related).

In [233]:
# Category
display_table(android_final, 1 )

FAMILY : 18.936914569461685
GAME : 9.694165444080802
TOOLS : 8.452770567656021
BUSINESS : 4.593161042771697
LIFESTYLE : 3.9047511567543167
PRODUCTIVITY : 3.8934657487868187
FINANCE : 3.7016138133393524
MEDICAL : 3.521047285859384
SPORTS : 3.3969077982169056
PERSONALIZATION : 3.317909942444419
COMMUNICATION : 3.2389120866719328
HEALTH_AND_FITNESS : 3.080916375126961
PHOTOGRAPHY : 2.9454914795169844
NEWS_AND_MAGAZINES : 2.7987811759395105
SOCIAL : 2.663356280329534
TRAVEL_AND_LOCAL : 2.3360794492720913
SHOPPING : 2.245796185532107
BOOKS_AND_REFERENCE : 2.144227513824625
DATING : 1.8620923146371742
VIDEO_PLAYERS : 1.794379866832186
MAPS_AND_NAVIGATION : 1.3993905879697552
FOOD_AND_DRINK : 1.2413948764247829
EDUCATION : 1.1736824286197944
ENTERTAINMENT : 0.9592596772373322
LIBRARIES_AND_DEMO : 0.9366888613023362
AUTO_AND_VEHICLES : 0.9254034533348381
HOUSE_AND_HOME : 0.8238347816273558
WEATHER : 0.8012639656923597
EVENTS : 0.7109807019523756
PARENTING : 0.6545536621148855
ART_AND_DESIGN : 

Google Play seems significantly different: There are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc). We can see that the family category (almost 19% of the apps) is mostly for kids.<br>

Practical apps seem to have a better representation on Google Play compare to Apple Store 

In [234]:
# Genres
display_table(android_final, -4)

Tools : 8.441485159688522
Entertainment : 6.071549486513937
Education : 5.349283376594064
Business : 4.593161042771697
Productivity : 3.8934657487868187
Lifestyle : 3.8934657487868187
Finance : 3.7016138133393524
Medical : 3.521047285859384
Sports : 3.464620246021894
Personalization : 3.317909942444419
Communication : 3.2389120866719328
Action : 3.103487191061957
Health & Fitness : 3.080916375126961
Photography : 2.9454914795169844
News & Magazines : 2.7987811759395105
Social : 2.663356280329534
Travel & Local : 2.3247940413045933
Shopping : 2.245796185532107
Books & Reference : 2.144227513824625
Simulation : 2.0426588421171425
Dating : 1.8620923146371742
Arcade : 1.8508069066696762
Video Players & Editors : 1.7718090508971898
Casual : 1.749238234962194
Maps & Navigation : 1.3993905879697552
Food & Drink : 1.2413948764247829
Puzzle : 1.1285407967498025
Racing : 0.9931159011398263
Role Playing : 0.9366888613023362
Libraries & Demo : 0.9366888613023362
Auto & Vehicles : 0.925403453334838

We can see that <i>Genre</i> it has more categories that the <i>Category</i> column. We are only looking fo the big picture at the moment, so we'll only work with Category column moving forward. <br>

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

## Most Popular Apps by Genre on the App Store
One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.<br>

Below, we calculate the average number of user ratings per app genre on the App Store:

In [237]:
# Generate a frequency table for ios_final dataset by using previous function created
genres_ios = freq_table(ios_final, -5)

# Iterate through each row of the table
for genre in genres_ios:
    total = 0 # will store the sum() of ratings, not the actual rating
    len_genre = 0 # will store number of apps specific to each genre
    # iterate over the ios_final dataset
    for app in ios_final:
        genre_app = app[-5] # contains the app genre
        if genre_app == genre:
            n_ratings = float(app[5]) # save the user rating as float
            total += n_ratings # sum all ratings
            len_genre += 1 # increment 1 everytime a genre is repited
    # get the average num. user rating
    avg_n_ratings = total /len_genre
    # proint the app genre and avg. user ratings
    print(genre, ':', avg_n_ratings)

Photo & Video : 28441.54375
Games : 22812.903311965812
Music : 57326.530303030304
Social Networking : 71548.34905660378
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


### Exploring Navigation Genre

On average, navigation apps have the highest number of users reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together

In [248]:
# Iteratae through ios_final, to print the apps contained in 'Navigation' genre
for app in ios_final:
    # if genre is 'Navigation', ... then print app and its rating count
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5])

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


### Exploring Social Networking Genre

In [252]:
# Iteratae through ios_final, to print the apps contained in 'Social Networking' genre
for app in ios_final:
    # if genre is 'Navigation', ... then print app and its rating count
    if app[-5] == 'Social Networking':
        print(app[1], ':', app[5])

Facebook : 2974676
Pinterest : 1061624
Skype for iPhone : 373519
Messenger : 351466
Tumblr : 334293
WhatsApp Messenger : 287589
Kik : 260965
ooVoo – Free Video Call, Text and Voice : 177501
TextNow - Unlimited Text + Calls : 164963
Viber Messenger – Text & Call : 164249
Followers - Social Analytics For Instagram : 112778
MeetMe - Chat and Meet New People : 97072
We Heart It - Fashion, wallpapers, quotes, tattoos : 90414
InsTrack for Instagram - Analytics Plus More : 85535
Tango - Free Video Call, Voice and Chat : 75412
LinkedIn : 71856
Match™ - #1 Dating App. : 60659
Skype for iPad : 60163
POF - Best Dating App for Conversations : 52642
Timehop : 49510
Find My Family, Friends & iPhone - Life360 Locator : 43877
Whisper - Share, Express, Meet : 39819
Hangouts : 36404
LINE PLAY - Your Avatar World : 34677
WeChat : 34584
Badoo - Meet New People, Chat, Socialize. : 34428
Followers + for Instagram - Follower Analytics : 28633
GroupMe : 28260
Marco Polo Video Walkie Talkie : 27662
Miitomo : 2

The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:


In [250]:
# Iterate over the ios_final to print which apps contained in 'Reference' genre 
for app in ios_final:
    # if genre is 'Reference'... then print app name and its rating count
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

   - Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

   - Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

   - Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

Now let's analyze the Google Play market a bit.

## Most Popular Apps by Genre on Google Play
For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [267]:
# Call display_table() function for android_final to return frequency table

display_table(android_final, 5) # idx = 5 is the Installs columns

1,000,000+ : 15.743144114659746
100,000+ : 11.556257758717978
10,000,000+ : 10.518000225708159
10,000+ : 10.190723394650716
1,000+ : 8.396343527818532
100+ : 6.91795508407629
5,000,000+ : 6.838957228303803
500,000+ : 5.574991535944025
50,000+ : 4.773727570251665
5,000+ : 4.51416318699921
10+ : 3.5436181017943795
500+ : 3.250197494639431
50,000,000+ : 2.290937817402099
100,000,000+ : 2.1216566978896285
50+ : 1.9185193544746644
5+ : 0.7899785577248618
1+ : 0.5078433585374111
500,000,000+ : 0.27084979121995256
1,000,000,000+ : 0.22570815934996052
0+ : 0.0451416318699921
0 : 0.011285407967498025


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (**category**).

In [255]:
categories_android = freq_table(android_final, 1)

# iterate through the android dataset
for category in categories_android:
    total = 0 # will store the sum of installs specific to each genre
    len_genre = 0 # will store number of apps specific to each genre
    
    # iterate through android final
    for app in android_final:
        category_app = app[1] # save the app genre to variable called category_app
        if category_app == category:
            n_installs = app[5] # idx 5 it's located the num. of installs
            n_installs = n_installs.replace(',', '') # remove commas and +symbols
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_genre += 1
            
    # variable contains average installs per genre
    avg_n_installs = total/len_genre
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1820673.076923077
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15560965.599534342
FAMILY : 3694276.334922527
MEDICAL : 120616.48717948717
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17805627.643678162
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10682301.033377837
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

### Explore COMMUNICATION category apps...
Let's check which apps are in the COMMUNICATION category that has more than 100 or 500 millions installs

In [257]:
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+' 
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        
        # print app and its num. of installs
        print(app[0], ':', app[5])
                                      
        

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If we removed all the communications apps that have over 100 millions installs, the average would be reduced roughly ten times:

In [259]:
# variable will contain list of apps that have less than 100 millions installs
under_100_m = []

# iterate through android_final to find the average number of app under 100 mil.
for app in android_final:
    n_installs =  app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    # if category COMMUNICATION and n_installs less than 100 mil, then add to under_100_m
    if app[1] == 'COMMUNICATION' and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
# add all the under_100_m apps
sum(under_100_m)/ len(under_100_m)

3603485.3884615386



We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:


In [262]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [264]:
# Let's see these apps that are skewing the average
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [266]:
# iterate through android_final to find apps that fit in middle range of popularity
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                           or app[5] == '5,000,000+'
                                           or app[5] == '10,000,000+'
                                           or app[5] == '50,000,000+'):
        
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

## Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.


### Second Part: Study the Genre for Androind Apps

In [281]:
genre_android = freq_table(android_final, 9)

for genre in genre_android:
    total = 0
    len_genre = 0
    for app in android_final:
        genre_app = app[9]
        if genre_app == genre:
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_genre += 1
            
    avg_n_installs = total / len_genre
    print(genre, ':', avg_n_installs)
            

Art & Design : 2122850.9433962265
Art & Design;Creativity : 285000.0
Auto & Vehicles : 647317.8170731707
Beauty : 513151.88679245283
Books & Reference : 8767811.894736841
Business : 1712290.1474201474
Comics : 831873.1481481482
Comics;Creativity : 50000.0
Communication : 38456119.167247385
Dating : 854028.8303030303
Education : 540691.7721518987
Education;Creativity : 2875000.0
Education;Education : 4759517.0
Education;Pretend Play : 1800000.0
Education;Brain Games : 5333333.333333333
Entertainment : 5602792.775092937
Entertainment;Brain Games : 3314285.714285714
Entertainment;Creativity : 4000000.0
Entertainment;Music & Video : 6413333.333333333
Events : 253542.22222222222
Finance : 1387692.475609756
Food & Drink : 1924897.7363636363
Health & Fitness : 4188821.9853479853
House & Home : 1331540.5616438356
Libraries & Demo : 638503.734939759
Lifestyle : 1412998.3449275363
Lifestyle;Pretend Play : 10000000.0
Arcade : 22888365.48780488
Puzzle : 8302861.91
Racing : 15910645.681818182
Sport

## Aproach for Non-Free Apps